File: leetcode/algorithms/longest_valid_parentheses/longest_valid_parentheses.cc
class Solution {
public:
    int longestValidParentheses(string s) {
        int longest_substring_length = 0;
        int n = s.size();

        vector<int> lengths(n, 0);

        for (int i = 1; i < n; ++i) {
            if (s[i] == ')') {
                if (s[i - 1] == '(') {
                    lengths[i] = 2;
                    if (i >= 2) {
                        lengths[i] += lengths[i - 2];
                    }
                } else {
                    int length = lengths[i - 1];
                    int j = i - length - 1;

                    if (j < 0 || s[j] != '(') {
                        continue;
                    }

                    lengths[i] = lengths[i - 1] + 2;

                    if (j > 0) {
                        lengths[i] += lengths[j - 1];
                    }
                }
            }
            longest_substring_length = max(longest_substring_length, lengths[i]);
        }

        return longest_substring_length;
    }
};

File: leetcode/algorithms/longest_valid_parentheses/longest_valid_parentheses.py
class Solution(object):
    def longestValidParentheses(self, s):
        longest_substring_length = 0
        n = len(s)

        lengths = [0] * n

        for i, c in enumerate(s[1:], 1):
            if c == ')':
                if s[i - 1] == '(':
                    lengths[i] = 2
                    if i >= 2:
                        lengths[i] += lengths[i - 2]
                else:
                    length = lengths[i - 1]
                    j = i - length - 1

                    if j < 0 or s[j] != '(':
                        continue

                    lengths[i] = lengths[i - 1] + 2

                    if j > 0:
                        lengths[i] += lengths[j - 1]

            longest_substring_length = max(longest_substring_length, lengths[i])

        return longest_substring_length
